pkg/utils/cloud/gcp/iam.go

```
@@ -15,3 +15,63 @@
  */
 
 package gcp
+
+import (
+	"context"
+
+	"github.com/appvia/kore/pkg/utils"
+
+	resourcemanager "google.golang.org/api/cloudresourcemanager/v1beta1"
+	"google.golang.org/api/option"
+)
+
+// MaxChunkSize is the largest number of permissions that can be checked in one request
+const MaxChunkSize = 100
+
+// CheckServiceAccountPermissions is responsible for checking the service account the permissions
+func CheckServiceAccountPermissions(
+	ctx context.Context,
+	project string,
+	serviceAccountKey string,
+	list []string,
+) (bool, []string, error) {
+	options := option.WithCredentialsJSON([]byte(serviceAccountKey))
+
+	// @step: we create a client from the service account first
+	client, err := resourcemanager.NewService(context.Background(), options)
+	if err != nil {
+		return false, nil, err
+	}
+
+	return HasPermissions(ctx, project, client, list)
+}
+
+// HasPermissions checks if we have the correct permissions
+func HasPermissions(
+	ctx context.Context,
+	project string,
+	client *resourcemanager.Service,
+	list []string) (bool, []string, error) {
+
+	var missing []string
+
+	for _, chunk := range utils.ChunkBy(list, MaxChunkSize) {
+		request := &resourcemanager.TestIamPermissionsRequest{
+			Permissions: chunk,
+		}
+
+		resp, err := client.Projects.TestIamPermissions(project, request).Context(ctx).Do()
+		if err != nil {
+			return false, nil, err
+		}
+		if len(resp.Permissions) != len(chunk) {
+			for _, x := range chunk {
+				if !utils.Contains(x, resp.Permissions) {
+					missing = append(missing, x)
+				}
+			}
+		}
+	}
+
+	return (len(missing) <= 0), missing, nil
```

I would return the error here instead of return here the missing ones

---

So i've split into two methods, one which provides with a yes or no and error and one which provides the same but with the missing perms https://github.com/appvia/kore/pull/368/commits/68c99f107fb8b52dfdb3c6e811cd58dc8579c606